src/crimson/osd/pg_map.h

```diff
@@ -10,12 +10,75 @@
 
 #include "include/types.h"
 #include "crimson/common/type_helpers.h"
+#include "crimson/common/smp_helpers.h"
 #include "crimson/osd/osd_operation.h"
 #include "osd/osd_types.h"
 
 namespace crimson::osd {
 class PG;
 
+/**
+ * PGShardMapping
+ *
+ * Maps pgs to shards.
+ */
+class PGShardMapping {
+public:
+  /// Returns mapping if present, NULL_CORE otherwise
+  core_id_t get_pg_mapping(spg_t pgid) {
+    auto iter = pg_to_core.find(pgid);
+    ceph_assert(iter == pg_to_core.end() || iter->second != NULL_CORE);
```

Comment:
Any reason for `ceph_assert()` instead of `assert()`?

Reply:
Because I don't want it removed in release builds.

Reply:
I changed it in this file, but I'll leave adjusting the others to later.  Btw, ceph_assert has historically meant, in this code base, that you don't want it removed in release builds and there's a ton of code written with that assumption.  What is the value in trying to change it?

Reply:
Mmm, ceph_assert_always has been around for a while, I'll try to use it in the future.

Reply:
The story with `ceph_assert` and `ceph_assert_always` is really long. IIRC a few years back we expressed the willingness to convert all `ceph_asserts` into plain `asserts`. The  _always_ variant was introduced then to let preserve assertions even in production builds.